### PR TITLE
casper: Case 1 for main_hash_voted_frac. So far error at z3 query.

### PR DESCRIPTION
**Pull request by denis-bogdanas**

### CODE REVIEW

**Comment by daejunpark** [comment]
@denis-bogdanas Please use `;` for the comments for INI files.

**Comment by denis-bogdanas** [comment]
Can you merge this pull request? At the moment main_hash_voted_frac() and total_curdyn_deposits_scaled() are completed and updated to latest Casper/EVM.

File: casper/casper-spec.ini [hunk 1]
```diff
@@ -44,6 +44,128 @@ requires:
...
+    andBool 0 <=Int CURRENT_EPOCH           andBool CURRENT_EPOCH           <=Int maxSInt128
+    andBool 0  <Int CUR_EPOCH_SCALE_FACTOR  andBool CUR_EPOCH_SCALE_FACTOR  <=Int maxDecimal
+    andBool TOTAL_CURDYN_DEPOSITS *Int CUR_EPOCH_SCALE_FACTOR               <Int pow255
+    andBool DECIMAL_RESULT ==Int TOTAL_CURDYN_DEPOSITS *Int CUR_EPOCH_SCALE_FACTOR /Int VYPER_DEC_MULTIPLIER
+    andBool DECIMAL_RESULT <=Int maxDecimal
+    andBool RESULT ==Int DECIMAL_RESULT /Int VYPER_DEC_MULTIPLIER
+ensures:
+    ensures 0 <=Int RESULT                  andBool RESULT <=Int maxSInt128
+attribute:
+
```
Comment (daejunpark) [inline casper/casper-spec.ini]: @denis-bogdanas how about the negative cases for `total_curdyn_deposits_scaled`?

File: Makefile [hunk 2]
```diff
@@ -210,31 +214,10 @@ $(specs_dir)/examples/sum-to-n-spec.k: resources/sum-to-n.md $(TANGLER)
...
-	mkdir -p $(dir $@)
-	python3 resources/gen-spec.py $^ vote recommended_target_hash vote > $@
-	cp casper/abstract-semantics.k $(dir $@)
-	cp casper/verification.k $(dir $@)
-
-$(specs_dir)/casper/delete_validator-spec.k: $(casper_tmpls) casper/casper-spec.ini
-	@echo >&2 "==  gen-spec: $@"
-	mkdir -p $(dir $@)
-	python3 resources/gen-spec.py $^ delete_validator delete_validator > $@
+	python3 resources/gen-spec.py $^ $* $* > $@
```
Comment (yzhang90) [inline Makefile]: It will not work for `vote` spec, because `vote` requires `recommended_target_hash` and `proc_reward`. You can not use `$*`.

**Comment by denis-bogdanas** [comment]
But `recommended_target_hash-spec.k` and `proc_reward-spec.k` are also created.

**Comment by daejunpark** [comment]
Note that the vote spec file (`vote-spec.k`) should contain the other two function's specs in the *same* file.

**Comment by denis-bogdanas** [comment]
Ok, I'll see how to fix this makefile issue.

**Comment by denis-bogdanas** [comment]
Can you do a code review now? `total_curdyn_deposits_scaled()` is completely proved. I also fixed the `Makefile`. Should satisfy requirements for `vote-spec.k`.

**Comment by denis-bogdanas** [comment]
I forgot reverting the code on last line. Please revert it.

**Comment by denis-bogdanas** [comment]
Added all public constant functions and increment_dynasty().

**Comment by daejunpark** [comment]
@denis-bogdanas I cannot reproduce the result. Which versions of KEVM and K are you using? Also, what changes made to the bytecode?

File: casper/verification.k [hunk 3]
```diff
@@ -66,13 +66,30 @@ module VERIFICATION
...
+      requires A >=Int 0 andBool B >=Int 0 andBool A =/=Int 0 andBool A *Int B <=Int maxSInt256
+
+    rule A >Int maxUInt256 => true
+      requires A >Int maxSInt256 andBool chop(A) <=Int maxSInt256
+
+    rule chop(A *Int B) /Int A ==K B => false
+      requires A *Int B >Int maxUInt256
+
+    rule 0 -Word ((pow256 -Int chop( A *Int B )) /Int A) ==K B => false
+      requires A >=Int 0 andBool B >=Int 0 andBool chop(A *Int B) >Int maxSInt256
```
Comment (daejunpark) [inline casper/verification.k]: could you explain a bit about this? it's not trivial to me.

**Comment by denis-bogdanas** [comment]
I just added proofs to the 2 lemmas in comments. Also added some requires clauses that were always true in evm context, but have to be here for correctness.

**Comment by denis-bogdanas** [comment]
@daejunpark  Verified on latest versions of K and kevm that you suggested.

**Pull request merged**